app/modules/scheduling/__init__.py

```
@@ -0,0 +1,135 @@
+
+def _replace_function_source(function, replacements: list[tuple[str, str]]):
+    """Recompile one existing scheduling function with narrowly-scoped fixes."""
+    source = inspect.getsource(function)
+    for old, new in replacements:
+        if old not in source:
+            raise RuntimeError(
+                f"Scheduling compatibility patch could not find expected source in {function.__name__}."
+            )
+        source = source.replace(old, new, 1)
+    namespace = _solver.__dict__
+    exec(compile(source, inspect.getsourcefile(function) or "<scheduling>", "exec"), namespace, namespace)
```

Ruff (0.16.1)

[error] Use of `exec` detected

(S102)